Migrate platform reference to Upbound official providers

ytsarev opened the pull request:
<!--
Thank you for helping to improve Upbound!

Please read through https://git.io/fj2m9 if this is your first time opening an
Upbound pull request. Find us in https://slack.crossplane.io/messages/upbound if
you need any help contributing.
-->

### Description of your changes

* Migrate to official provider-gcp
* Include CompositeNetwork XR within the Cluster Composition in a consistent manner similar to aws and azure platform-refs
* Comply with new GKE authentication requirements and extend abstractions with googleApplicationCredentialsSecret

I have:

- [x] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR, as appropriate.

### How has this code been tested

```
kubectl apply -f examples/cluster.yaml
```

```
k get managed
NAME                                                              CHART                   VERSION   SYNCED   READY   STATE      REVISION   DESCRIPTION        AGE
release.helm.crossplane.io/platform-ref-gcp-cluster-jvl4j-nb5fw   kube-prometheus-stack   34.5.1    True     True    deployed   1          Install complete   27m

NAME                                                                     READY   SYNCED   EXTERNAL-NAME                                                                                                                       AGE
projectiammember.cloudplatform.gcp.upbound.io/platform-ref-gcp-cluster   True    True     crossplane-playground/roles/container.admin/serviceAccount:platform-ref-gcp-cluster@crossplane-playground.iam.gserviceaccount.com   27m

NAME                                                                   READY   SYNCED   EXTERNAL-NAME              AGE
serviceaccount.cloudplatform.gcp.upbound.io/platform-ref-gcp-cluster   True    True     platform-ref-gcp-cluster   27m

NAME                                                                      READY   SYNCED   EXTERNAL-NAME                                                                                                                                                         AGE
serviceaccountkey.cloudplatform.gcp.upbound.io/platform-ref-gcp-cluster   True    True     projects/crossplane-playground/serviceAccounts/platform-ref-gcp-cluster@crossplane-playground.iam.gserviceaccount.com/keys/62cc39a87cca9bef38028f05bcd4614071461e12   27m

NAME                                                                     READY   SYNCED   EXTERNAL-NAME                          AGE
subnetwork.compute.gcp.upbound.io/platform-ref-gcp-cluster-jvl4j-mlr4l   True    True     platform-ref-gcp-cluster-jvl4j-mlr4l   27m

NAME                                                                  READY   SYNCED   EXTERNAL-NAME                          AGE
network.compute.gcp.upbound.io/platform-ref-gcp-cluster-jvl4j-vznfl   True    True     platform-ref-gcp-cluster-jvl4j-vznfl   27m

NAME                                                                     READY   SYNCED   EXTERNAL-NAME                          AGE
nodepool.container.gcp.upbound.io/platform-ref-gcp-cluster-jvl4j-nbnns   True    True     platform-ref-gcp-cluster-jvl4j-nbnns   27m

NAME                                                                    READY   SYNCED   EXTERNAL-NAME                          AGE
cluster.container.gcp.upbound.io/platform-ref-gcp-cluster-jvl4j-9j7g6   True    True     platform-ref-gcp-cluster-jvl4j-9j7g6   27m
```

```
k get composite
NAME                                                                  READY   COMPOSITION                      AGE
gke.gcp.platformref.upbound.io/platform-ref-gcp-cluster-jvl4j-5wvrp   True    gke.gcp.platformref.upbound.io   29m

NAME                                                                               READY   COMPOSITION                                        AGE
compositenetwork.gcp.platformref.upbound.io/platform-ref-gcp-cluster-jvl4j-w5q75   True    gcp.compositenetworks.gcp.platformref.upbound.io   29m

NAME                                                                         READY   COMPOSITION                                        AGE
compositecluster.gcp.platformref.upbound.io/platform-ref-gcp-cluster-jvl4j   True    gke.compositeclusters.gcp.platformref.upbound.io   29m

NAME                                                                       READY   COMPOSITION                           AGE
services.gcp.platformref.upbound.io/platform-ref-gcp-cluster-jvl4j-sjnk8   True    services.gcp.platformref.upbound.io   29m
```

muvaf commented on cluster/services/composition.yaml:
```diff
@@ -1,13 +1,13 @@
   writeConnectionSecretsToNamespace: upbound-system
   compositeTypeRef:
-    apiVersion: gcp.platformref.crossplane.io/v1alpha1
+    apiVersion: gcp.platformref.upbound.io/v1alpha1
```
I think this composition is missing a `ProviderConfig.helm` object.

ytsarev replied:
it's a part of GKE Composition

ytsarev commented on examples/network.yaml:
```diff
@@ -0,0 +1,8 @@
+spec:
+  id: platform-ref-gcp-cluster
+  clusterRef:
+   id: platform-ref-gcp-cluster
```
Thanks for the catch, it was leftover from previous ref state

ytsarev commented on examples/network.yaml:
```diff
@@ -0,0 +1,8 @@
+apiVersion: gcp.platformref.upbound.io/v1alpha1
```
following the strategy in ref-aws and ref-gcp, I renamed example file to `network-xr.yaml` and `cluster-claim.yaml` to reflect their kind

ytsarev commented on cluster/composition.yaml:
```diff
@@ -1,25 +1,29 @@
   writeConnectionSecretsToNamespace: upbound-system
   compositeTypeRef:
-    apiVersion: gcp.platformref.crossplane.io/v1alpha1
+    apiVersion: gcp.platformref.upbound.io/v1alpha1
```
`crossplane.yaml` was updated

ytsarev commented on cluster/gke/definition.yaml:
```diff
@@ -48,18 +48,26 @@ spec:
                     - count
                     - size
-                  networkRef:
+                  googleApplicationCredentialsSecret:
```
Added, it works great, thanks a lot for the suggestion

ytsarev commented on cluster/gke/composition.yaml:
```diff
@@ -1,39 +1,28 @@
 kind: Composition
 metadata:
-  name: gke.gcp.platformref.crossplane.io
+  name: gke.gcp.platformref.upbound.io
```
Composition extended, things to note:

* I had to add ServiceAccount management role to the account that is used for GCP provider credentials
* Currently I hit
```
  Warning  CannotUpdateManagedResource      3m8s (x111 over 8m36s)  managed/container.gcp.upbound.io/v1beta1, kind=cluster  Operation cannot be fulfilled on clusters.container.gcp.upbound.io "platform-ref-gcp-cluster-fd76b-xbn5d": the object has been modified; please apply your changes to the latest version and try again
```
Maybe it is due 
```
              nodeConfig:
                - serviceAccountSelector:
                    matchControllerRef: true
```
and similar to https://github.com/upbound/official-providers/issues/466#issuecomment-1253444978 , though the selector is part of the `spec` not a `patch` there, so I do not yet properly understand the root cause

@muvaf should we propagate ServiceAccount ID through XR status here?

muvaf replied:
@ytsarev Yes, I'm afraid we need to patch through status. It's an array and core crossplane replaces the array item when you try to patch, hence it ends up having only `matchControllerRef`.

muvaf commented on crossplane.yaml:
```diff
@@ -34,65 +33,11 @@ metadata:
+    - provider: xpkg.upbound.io/upbound/provider-gcp
+      version: ">=v0.13.0-0"
+    - provider: xpkg.upbound.io/crossplane/provider-helm
+      version: ">=v0.10.0-0"
```
```suggestion
      version: ">=v0.11.1-0"
```

ytsarev replied:
@muvaf , thanks, updated!

During the configuration testing, I got v0.12.rc build installed. Is that expected?
```
 k get providers
NAME                       INSTALLED   HEALTHY   PACKAGE                                                            AGE
provider-aws               True        True      xpkg.upbound.io/upbound/provider-aws:v0.14.0                       5d15h
provider-gcp               True        True      xpkg.upbound.io/upbound/provider-gcp:v0.13.0                       16h
crossplane-provider-helm   True        True      xpkg.upbound.io/crossplane/provider-helm:v0.12.0-rc.0.5.gb368192   30s
```

ytsarev commented:
@muvaf @stevendborrelli this PR is ready for the test, Configuration is also installable via examples/configuration.yaml
```
apiVersion: pkg.crossplane.io/v1
kind: Configuration
metadata:
  name: platform-ref-gcp-staging
spec:
  package: xpkg.upbound.io/upbound/platform-ref-gcp-staging:v0.1.1
  packagePullSecrets:
```

muvaf commented on cluster/gke/composition.yaml:
```diff
@@ -1,39 +1,96 @@
+        kind: ServiceAccount
+      patches:
+        - fromFieldPath: spec.id
+          toFieldPath: metadata.name
```
```suggestion
          toFieldPath: metadata.annotations[crossplane.io/external-name]
```
This is a better way of naming resources in general so that Crossplane can take care of `metadata.name`, making sure there is no conflict.

ytsarev replied:
Ah, got it. I always had an opposite vector of `name -> external-name` in mind. That makes sense, will update

muvaf commented on cluster/gke/composition.yaml:
```diff
@@ -1,39 +1,96 @@
+            serviceAccountIdSelector:
+              matchControllerRef: true
+      patches:
+        - fromFieldPath: spec.id
+          toFieldPath: metadata.name
```
```suggestion
```

Seems like keys don't have names. Not patching it at all would be enough here to prevent conflicts I think.

muvaf commented on examples/cluster-claim.yaml:
```diff
@@ -14,8 +14,6 @@ spec:
     services:
```
I think using `1` as count would be faster to provision.

muvaf commented on examples/configuration.yaml:
```diff
@@ -0,0 +1,8 @@
+metadata:
+  name: platform-ref-gcp-staging
+spec:
+  package: xpkg.upbound.io/upbound/platform-ref-gcp-staging:v0.1.2
```
Don't forget to replace with the real repo before merging.

muvaf reviewed (approved):
Tested it a couple of times. Everything works as expected. For some reason, the cluster goes into repairing state first during deletion which makes it take longer to delete but that's likely about GCP API. Thanks a lot @ytsarev !!

donovanmuller commented:
> Tested it a couple of times. Everything works as expected. For some reason, the cluster goes into repairing state first during deletion which makes it take longer to delete but that's likely about GCP API. Thanks a lot @ytsarev !!

@muvaf I actually ran into something similar [here](https://github.com/upbound/official-providers/pull/657#:~:text=Had%20to%20split%20container/Cluster%20and%20NodePool%20examples%20because%20of%20an%20issue%20related%20to%20adding%20two%20node%20pools.%20See%20this) as part of https://github.com/upbound/official-providers/pull/657, where if there are two node pools on creation, it causes this behaviour. See [this](https://www.reddit.com/r/googlecloud/comments/gewqaq/comment/g63g98e/?utm_source=reddit&utm_medium=web2x&context=3) related post.

muvaf commented:
Test with latest state. The deletion succeeded, too.
```
NAME                                                              CHART                   VERSION   SYNCED   READY   STATE      REVISION   DESCRIPTION        AGE
release.helm.crossplane.io/platform-ref-gcp-cluster-4rdlm-mchl6   kube-prometheus-stack   34.5.1    True     True    deployed   1          Install complete   12m

NAME                                                                                 READY   SYNCED   EXTERNAL-NAME                                                                                                                       AGE
projectiammember.cloudplatform.gcp.upbound.io/platform-ref-gcp-cluster-4rdlm-xfqb8   True    True     crossplane-playground/roles/container.admin/serviceAccount:platform-ref-gcp-cluster@crossplane-playground.iam.gserviceaccount.com   12m

NAME                                                                               READY   SYNCED   EXTERNAL-NAME              AGE
serviceaccount.cloudplatform.gcp.upbound.io/platform-ref-gcp-cluster-4rdlm-mwz2d   True    True     platform-ref-gcp-cluster   12m

NAME                                                                                  READY   SYNCED   EXTERNAL-NAME                                                                                                                                                         AGE
serviceaccountkey.cloudplatform.gcp.upbound.io/platform-ref-gcp-cluster-4rdlm-5xb2p   True    True     projects/crossplane-playground/serviceAccounts/platform-ref-gcp-cluster@crossplane-playground.iam.gserviceaccount.com/keys/33ef64d7fd28e5c02fc3e609ae3d3ffedd96b0fe   12m

NAME                                                                  READY   SYNCED   EXTERNAL-NAME                          AGE
network.compute.gcp.upbound.io/platform-ref-gcp-cluster-4rdlm-7gcsm   True    True     platform-ref-gcp-cluster-4rdlm-7gcsm   12m

NAME                                                                     READY   SYNCED   EXTERNAL-NAME                          AGE
subnetwork.compute.gcp.upbound.io/platform-ref-gcp-cluster-4rdlm-btdwh   True    True     platform-ref-gcp-cluster-4rdlm-btdwh   12m

NAME                                                                     READY   SYNCED   EXTERNAL-NAME                          AGE
nodepool.container.gcp.upbound.io/platform-ref-gcp-cluster-4rdlm-2lj59   True    True     platform-ref-gcp-cluster-4rdlm-2lj59   12m

NAME                                                                    READY   SYNCED   EXTERNAL-NAME                          AGE
cluster.container.gcp.upbound.io/platform-ref-gcp-cluster-4rdlm-jcxbd   True    True     platform-ref-gcp-cluster-4rdlm-jcxbd   12m
```

jbw976 commented on README.md:
```diff
@@ -95,13 +95,16 @@ kubectl get pkg
 * `roles/compute.networkAdmin`
 * `roles/container.admin`
 * `roles/iam.serviceAccountUser`
+* `roles/iam.securityAdmin`
```
what caused these extra roles/permissions to now be needed?

ytsarev replied:
@jbw976 it is required to create `ProjectIAMMember`

I mentioned here https://github.com/upbound/platform-ref-gcp/pull/22#issuecomment-1259399923

we need all that machinery to enable automated access in the form of

```
          identity:
            type: GoogleApplicationCredentials
```
for the helm ProviderConfig

jbw976 replied:
aha yes, thanks for clearing that up.  So just to tie some loose ends together (for myself), this is the new approach for getting Auth to work with GKE clusters.  This support was added to:

* provider-helm in https://github.com/crossplane-contrib/provider-helm/pull/109
* provider-kubernetes in https://github.com/crossplane-contrib/provider-kubernetes/pull/10

And that functionality generally address issues we've seen around this topic, like:

* https://github.com/crossplane-contrib/provider-gcp/issues/343
* https://github.com/crossplane-contrib/provider-gcp/issues/415

jbw976 commented on README.md:
```diff
@@ -240,12 +243,12 @@ kubectl crossplane install configuration ${PLATFORM_CONFIG}
 
 The cloud service primitives that can be used in a `Composition` today are
-listed in the Crossplane provider docs:
+listed in the Marketplace provider docs:
```
"Upbound Marketplace" instead of just "Marketplace"?

jbw976 reviewed (commented):
Thanks for another quick but thorough investment into our major reference platforms @ytsarev!! 💪 

A few questions to explore, but mostly looks good - an approval isn't far away 😂

ytsarev commented on cluster/composition.yaml:
```diff
@@ -32,10 +36,13 @@ spec:
-          toFieldPath: spec.parameters.networkRef.id
+        - fromFieldPath: spec.id
+          toFieldPath: spec.parameters.compositeNetworkSelector.matchLabels[networks.gcp.platformref.upbound.io/network-id]
+        - fromFieldPath: spec.parameters.googleApplicationCredentialsSecret.name
+        - fromFieldPath: spec.parameters.googleApplicationCredentialsSecret.namespace
+        - fromFieldPath: spec.parameters.googleApplicationCredentialsSecret.key
```
@jbw976 without `toFieldPath` it is patching exactly the same destination as mentioned in `fromFieldPath`

But these patches are actually leftovers from the previous intermediate implementation so I remove them anyway, thanks for the catch!

ytsarev commented on cluster/network/definition.yaml:
```diff
@@ -0,0 +1,25 @@
+apiVersion: apiextensions.crossplane.io/v1
```
@jbw976 that's totally right, this PR makes network XR layout consistent with  aws and azure refs:
* Use Network XR from Cluster Composition
* Do not provide Claim, just XR as [example](https://github.com/upbound/platform-ref-gcp/pull/22/files#diff-3d629f71a3fea43b9887117b57216cf2fbcdfde102ed609bc9ae17bb54fd4a5e) so it also appears in the Marketplace

jbw976 commented:
Do you want to squash the commit history a bit @ytsarev?

ytsarev commented:
@jbw976 I've squashed the commits into one leaving only useful messages from the commit log. I've just updated the Configuration example to point to the semantical `v0.1.0` version as we agreed with @muvaf .

The PR is good to merge, thank you so much for the thorough review, folks!
